Add e2e test utils to install and activate themes

johnwatkins0 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

This fixes #23681 by adding two new test utils, `installTheme` and `activateTheme`. As far as I know these are not directly useful right now in the Gutenberg plugin, but I have had need for them on other projects that use `e2e-test-utils` (and I believe they could be useful in this plugin in the future). 

**Update:** Also added `deleteTheme` and `themeInstalled` utils, along with a theme switching e2e test that installs and activates a theme, creates a post to confirm the editor works, and then deletes the theme.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Built using a local copy of the plugin. Imported these functions into a separate project and successfully used them in e2e tests.

**Update:** Also added an e2e test at @gziolo's suggestion.

## Example

```JS
import { installTheme, activateTheme } from '@wordpress/e2e-test-utils';

describe( 'My plugin test', () => {
	it( 'does something different when a different theme is active', async () => {
		await installTheme( 'twentynineteen' );
		await activateTheme( 'twentynineteen' );

		await expect( page ).not.toPassSomeTest();

		await installTheme( 'twentytwenty' );
		await activateTheme( 'twentytwenty' );

		await expect( page ).toPassSomeTest();
	} );
} );

```

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

Nobreaking change introducing new test utils.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

johnwatkins0 committed:
(#23681) add e2e test utils to install and activate themes

youknowriad commented on packages/e2e-test-utils/src/activate-theme.js:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @param {string} slug Theme slug.
+ */
+export async function activateTheme( slug ) {
```
Do you think we need functions to deactivate and uninstall themes as well? (like plugins).

johnwatkins0 replied:
I can see uninstalling themes being useful. I'll add that. As for deactivating, deactivating a theme just means switching to a different theme, so I think that's covered by `activateTheme`.

johnwatkins0 replied:
Done in aef01296553bf7b3ac75193ada82d5f4b6996b5a

gziolo commented:
Nice 👍 Can you add a simple test for Gutenberg that switches the theme to something else than the default one for a duration of this test? Something very basic that verifies the block editor still works? This way we will be able to ensure that new test helpers work properly.

johnwatkins0 committed:
Add deleteTheme e2e test util

johnwatkins0 committed:
Add themeInstalled util

johnwatkins0 committed:
Updates and docs for theme installation test utils

johnwatkins0 committed:
Add tests for theme switching

johnwatkins0 commented:
I need to figure out why theme activation is timing out in the test environment. 

> TimeoutError: waiting for selector ".theme[data-slug="primer"] .activate" failed: timeout 30000ms exceeded

youknowriad commented on packages/e2e-test-utils/src/theme-installed.js:
```diff
@@ -0,0 +1,23 @@
+ * @param {string} slug Theme slug to check.
+ * @return {boolean} Whether the theme exists.
+ */
+export async function themeInstalled( slug ) {
```
Should this be isThemeInstalled?

youknowriad replied:
This is the last blocker for me. themeInstalled is not very clear IMO.

youknowriad commented on packages/e2e-tests/specs/editor/various/theme-switching.test.js:
```diff
@@ -0,0 +1,50 @@
+// These can be any themes available from wordpress.org that don't themselves cause editor errors.
+const testThemes = [ [ 'primer' ], [ 'astra' ] ];
+
+describe( 'Theme switching', () => {
```
I know it was @gziolo's suggestion but i'm personally not entirely sold on the idea of this test as it relies on a external API that can break depending on the internet connection... (We don't have the same thing for plugins)

That said, we can give it a try and see after some time whether it's stable enough...

youknowriad replied:
Also, this would be better a test in Core e2e tests :)

johnwatkins0 replied:
Yeah, I admit I had trouble figuring out where best to put this test, and where I decided to put it felt arbitrary. And then indeed this test is currently timing out in the build (while it passes locally) for a reason that I will need to investigate before this PR can be merged. 

@gziolo As a compromise would you be okay with removing the part of this test that downloads external themes? We can still test switching between themes that exist in the environment (i.e. any `twenty*` theme). Backup option is just removing this test -- any issues with the utils would surface in test environments anyway.

gziolo replied:
I don't think I ever considered external themes here, it should be perfectly fine to use one of many bundled themes.

johnwatkins0 replied:
👍  I will update this week.

gziolo commented:
This PR needs to be rebased. I think @youknowriad added some of the logic in the meantime. As discussed, let's keep only tests that can work with bundled themes. Thank you for the work done so far 👍

johnwatkins0 committed:
Merge branch 'master' into feature/23681-e2e-theme-utils

johnwatkins0 committed:
Remove theme switching test reliant on external api

johnwatkins0 committed:
Change themeInstalled to isThemeInstalled

johnwatkins0 committed:
Rebuild docs

johnwatkins0 committed:
Merge remote-tracking branch 'origin/master' into feature/23681-e2e-theme-utils

johnwatkins0 commented:
Thanks for the review @youknowriad @gziolo. I've removed the flaky test and changed `themeInstalled` to `isThemeInstalled`. Let me know if you see anything else.

gziolo commented on packages/e2e-test-utils/src/index.js:
```diff
@@ -58,7 +61,7 @@ export { setPostContent } from './set-post-content';
 export { switchUserToAdmin } from './switch-user-to-admin';
 export { switchUserToTest } from './switch-user-to-test';
-export { activateTheme } from './themes';
+export { isThemeInstalled } from './theme-installed';
```
We need to delete `themes.js` file as it is no longer in use.

gziolo commented on packages/e2e-test-utils/src/delete-theme.js:
```diff
@@ -0,0 +1,47 @@
+ * @param {string?} newThemeSlug A theme to switch to if the theme to delete is active. Required if the theme to delete is active.
+ * @param {string?} newThemeSearchTerm A search term to use if the new theme is not findable by its slug.
+ */
+export async function deleteTheme( slug, newThemeSlug, newThemeSearchTerm ) {
```
How do you feel about updating the function signature to put all optional params inside a single param? 
```suggestion
export async function deleteTheme( slug, { newThemeSlug, newThemeSearchTerm } ) {
```

gziolo commented on packages/e2e-test-utils/src/install-theme.js:
```diff
@@ -0,0 +1,41 @@
+ * @param {string} slug        Theme slug.
+ * @param {string?} searchTerm Search term to use if the theme is not findable by its slug.
+ */
+export async function installTheme( slug, searchTerm ) {
```
Similar idea as with `deleteTheme` what if we make `searchTerm` part of options config.

```suggestion
export async function installTheme( slug, { searchTerm } ) {
```

What do you think?

gziolo reviewed (approved):
Great work on those e2e test utils. It has already turned out to be useful as `activateTheme` was implemented after this PR was udpated. The file I mentioned needs to be deleted before proceeding because of that.

I had some thoughts in regards to the definition of `deleteTheme` and `installTheme` which is something worth considering but I don't feel super strong about it.

johnwatkins0 committed:
Remove obsolete activateTheme function

johnwatkins0 committed:
Use objects for function parameters

youknowriad commented:
Personally, I'd have preferred if all the helpers are in a single `theme.js` file. It's very hard to navigate these utils when they are separate as a util per file. I prefer better if it's grouped by purpose.

gziolo commented:
We can refactor the file structure separately. I don't mind having less files now that the number of helpers has grown 👍
